services/matching.js

```diff
@@ -0,0 +1,156 @@
+   * This might differ from the HTTP API as we have combined
+   * all the properties that depend on the order of coordinates into
+   * one object for ease of use.
+   *
+   * @typedef {Object} MatchPath
+   * @property {Array<number>} coordinates - An array containing (longitude, latitude).
+   * @property {'unrestricted'|'curb'} [approach="unrestricted"] - Used to indicate how requested routes consider from which side of the road to approach a waypoint.
+   * @property {number} [radius=5] - A number in meters indicating the assumed precision of the used tracking device.
+   * @property {boolean} [isWaypoint=true] - Whether this coordinate is waypoint or not. Note! the first and last coordinates will always have to be true.
+   * @property {string} [waypointName] - Custom names for waypoint used for the arrival instruction in banners and voice instructions.
+   * @property {date} [timestamp] - Unix timestamp corresponding the coordinate.
+   */
+  config.matchPath.forEach(function(obj) {
+    matchPath.coordinates.push(obj.coordinates[0] + ',' + obj.coordinates[1]);
+
+    // isWaypoint
+    if (obj.hasOwnProperty('isWaypoint') && obj.isWaypoint != null) {
+      matchPath.isWaypoint.push(obj.isWaypoint);
+    } else {
+      matchPath.isWaypoint.push(true); // default value
+    }
+
+    ['approach', 'radius', 'waypointName', 'timestamp'].forEach(function(prop) {
+      if (obj.hasOwnProperty(prop) && obj[prop] != null) {
+        matchPath[prop].push(obj[prop]);
+      } else {
+        matchPath[prop].push('');
+      }
+    });
+  });
+
+  ['coordinates', 'approach', 'radius', 'waypointName', 'timestamp'].forEach(
+    function(prop) {
+      // avoid sending params which are all `;`
+      if (
+        matchPath[prop].every(function(value) {
+          return value == '';
+        })
+      ) {
+        delete matchPath[prop];
+      } else {
+        matchPath[prop] = matchPath[prop].join(';');
+      }
+    }
+  );
+
+  // the api requires the first and last items to be true.
+  matchPath.isWaypoint[0] = true;
+  matchPath.isWaypoint[matchPath.isWaypoint.length - 1] = true;
+
+  if (
+    matchPath.isWaypoint.every(function(value) {
+      return value == true;
+    })
+  ) {
+    delete matchPath.isWaypoint;
+  } else {
+    // the api requires the indexes to be sent
+    matchPath.isWaypoint = matchPath.isWaypoint
+      .map(function(val, i) {
+        return val == true ? i : '';
+      })
+      .join(';');
+  }
+
+  var query = {
+    annotations: config.annotations,
+    geometries: config.geometries,
+    language: config.language,
+    overview: config.overview,
+    steps: config.steps,
+    tidy: config.tidy,
+    approaches: matchPath.approach,
+    radiuses: matchPath.radius,
+    waypoints: matchPath.isWaypoint,
+    timestamps: matchPath.timestamp,
+    waypoint_names: matchPath.waypointName
+  };
+
+  return this.client.createRequest({
+    method: 'GET',
```

Comment:
Do you think we might want to use the `POST` method, in order to avoid problems with URLs getting too long? https://www.mapbox.com/api-documentation/#using-http-post

Reply:
I tried using the `post` method, but it seems to not work. The server doesn't seem to understand `form-data` post requests. 
When I try to add the header `"content-type", "application/x-www-form-urlencoded"`, the header shows up in the chrome network inspection tab, but the post request is still made with 
![image](https://user-images.githubusercontent.com/6966254/40712247-d8675390-641a-11e8-9721-0e71358cf1a6.png)

Any idea what I might be missing?

Reply:
Here is the code snippet I used
```javascript
tryServiceMethod('matching', 'getMatching', {"matchPath": [{"coordinates":[-117.1728265285492,32.71204416018209]},{"coordinates":[-117.17288821935652,32.712258556224]},{"coordinates":[-117.17293113470076,32.712443613445814]},{"coordinates":[-117.17292040586472,32.71256999376694]},{"coordinates":[-117.17298477888109,32.712603845608285]},{"coordinates":[-117.17314302921294,32.71259933203019]},{"coordinates":[-117.17334151268004,32.71254065549407]}]
},accessToken , console.log)
```

Reply:
If I am getting it correctly, we need to implement `form-urlencoded` option in `MapRequest`.